src/parcsr_ls/par_mgr_solve.c

```diff
@@ -984,10 +985,23 @@ hypre_MGRCycle( void              *mgr_vdata,
             if (Frelax_type[level] == 2)
             {
                /* Do F-relaxation using AMG */
-               fine_grid_solver_solve((mgr_data -> aff_solver)[fine_grid],
-                                      A_ff_array[fine_grid],
-                                      F_fine_array[coarse_grid],
-                                      U_fine_array[coarse_grid]);
+               if (level == 0)
+               {
+                  /* TODO (VPM): unify with the next block */
```

Comment:
I agree these two should be merged at some point.

Reply:
Yes, the idea is to remove `fine_grid_solver_solve` and `fine_grid_solver_setup` from the MGR's struct. But first, we need to extend the idea of `hypre_Solver` to other preconditioners in hypre...